File: api_sample/python/rest-list-cameras/test_rest_list_cameras.py
# Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
"""
Offline tests for rest_list_cameras.py. No network, no server needed.

Run from this folder:  pytest -v
"""

import argparse

import pytest

import rest_list_cameras as sample


# ---------------------------------------------------------------------------
# Test doubles
# ---------------------------------------------------------------------------

class FakeResponse:
    def __init__(self, status_code=200, json_data=None, text=""):
        self.status_code = status_code
        self._json = json_data
        self.text = text

    @property
    def ok(self):
        return self.status_code < 400

    def json(self):
        if self._json is None:
            raise ValueError("no json")
        return self._json


class FakeSession:
    """Serves queued responses per verb and records calls (incl. DELETE count)."""

    def __init__(self, post=None, get=None, delete=None):
        self.verify = None
        self._post, self._get, self._delete = post, get, delete
        self.post_url = self.post_json = None
        self.get_url = self.get_headers = None
        self.delete_url = None
        self.delete_calls = 0

    def post(self, url, json=None, timeout=None):
        self.post_url, self.post_json = url, json
        return self._post

    def get(self, url, headers=None, timeout=None):
        self.get_url, self.get_headers = url, headers
        return self._get

    def delete(self, url, headers=None, timeout=None):
        self.delete_url = url
        self.delete_calls += 1
        return self._delete


# ---------------------------------------------------------------------------
# login()
# ---------------------------------------------------------------------------

def test_login_posts_credentials_and_stores_token():
    session = FakeSession(post=FakeResponse(200, {"token": "abc123"}))
    client = sample.NxServerClient("https://srv:7001", "admin", "pw", session=session)

    token = client.login()

    assert token == "abc123"
    assert session.post_url == "https://srv:7001/rest/v4/login/sessions"
    assert session.post_json == {"username": "admin", "password": "pw",
                                 "setCookie": False}


def test_login_unauthorized_raises():
    session = FakeSession(post=FakeResponse(401, text="bad"))
    client = sample.NxServerClient("https://srv:7001", "admin", "pw", session=session)
    with pytest.raises(sample.AuthError):
        client.login()


def test_login_without_token_raises_apierror():
    session = FakeSession(post=FakeResponse(200, {"nope": 1}))
    client = sample.NxServerClient("https://srv:7001", "admin", "pw", session=session)
    with pytest.raises(sample.ApiError):
        client.login()


# ---------------------------------------------------------------------------
# list_cameras()
# ---------------------------------------------------------------------------

def test_list_cameras_plain_array():
    payload = [{"id": "c1", "name": "Lobby", "status": "Online", "model": "Axis"}]
    session = FakeSession(get=FakeResponse(200, payload))
    client = sample.NxServerClient("https://srv:7001", "admin", "pw", session=session)
    client.token = "abc123"

    cams = client.list_cameras()

    assert cams[0]["name"] == "Lobby"
    assert session.get_url == "https://srv:7001/rest/v4/devices"
    assert session.get_headers["Authorization"] == "Bearer abc123"


def test_list_cameras_unwraps_reply_envelope():
    # Some Nx versions return {"reply": [...]}; the client should unwrap it.
    payload = {"reply": [{"id": "c1", "name": "Lobby"}]}
    session = FakeSession(get=FakeResponse(200, payload))
    client = sample.NxServerClient("https://srv:7001", "admin", "pw", session=session)
    client.token = "abc123"

    cams = client.list_cameras()

    assert len(cams) == 1
    assert cams[0]["name"] == "Lobby"


def test_list_cameras_without_login_raises():
    client = sample.NxServerClient("https://srv:7001", "admin", "pw",
                                   session=FakeSession())
    with pytest.raises(sample.ApiError):
        client.list_cameras()


# ---------------------------------------------------------------------------
# logout()
# ---------------------------------------------------------------------------

def test_logout_deletes_session_and_clears_token():
    session = FakeSession(delete=FakeResponse(200, {}))
    client = sample.NxServerClient("https://srv:7001", "admin", "pw", session=session)
    client.token = "abc123"

    client.logout()

    assert session.delete_calls == 1
    assert session.delete_url == "https://srv:7001/rest/v4/login/sessions/abc123"
    assert client.token is None


def test_logout_without_token_is_noop():
    session = FakeSession()
    client = sample.NxServerClient("https://srv:7001", "admin", "pw", session=session)
    client.logout()  # should not raise or call delete
    assert session.delete_calls == 0


# ---------------------------------------------------------------------------
# config + table
# ---------------------------------------------------------------------------

def test_config_uses_server_env_vars(monkeypatch):
    monkeypatch.setenv("NX_SERVER_HOST", "https://env:7001")
    args = argparse.Namespace(host=None, user=None, password=None)
    config = sample.resolve_config(args, {"NX_SERVER_HOST": "https://file:7001"})
    assert config["host"] == "https://env:7001"  # env beats file


def test_format_cameras_table():
    out = sample.format_cameras_table(
        [{"id": "c1", "name": "Lobby", "status": "Online", "model": "Axis"}])
    assert "NAME" in out and "Lobby" in out
    assert "No cameras" in sample.format_cameras_table([])

File: api_sample/node_js/rest-list-cameras/test_rest_list_cameras.mjs
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
/**
 * Offline tests for rest_list_cameras.mjs. No network, no server needed.
 *
 * Run from this folder:  node --test test_rest_list_cameras.mjs
 */

import assert from "node:assert/strict";
import test from "node:test";

import {
  NxServerClient,
  formatCamerasTable,
  resolveConfig,
  AuthError,
  ApiError,
} from "./rest_list_cameras.mjs";

function makeResponse({ status = 200, json = null, text = "" } = {}) {
  return {
    status,
    ok: status < 400,
    async json() {
      if (json === null) throw new Error("no json");
      return json;
    },
    async text() {
      return text;
    },
  };
}

/** Serves one queued response per verb; records calls (incl. DELETE count). */
function fakeFetch({ post = null, get = null, del = null } = {}) {
  const calls = { post: null, get: null, deleteUrl: null, deleteCalls: 0 };
  const impl = async (url, options = {}) => {
    const method = (options.method || "GET").toUpperCase();
    if (method === "POST") {
      calls.post = { url, body: options.body ? JSON.parse(options.body) : null };
      return post;
    }
    if (method === "DELETE") {
      calls.deleteUrl = url;
      calls.deleteCalls += 1;
      return del;
    }
    calls.get = { url, headers: options.headers };
    return get;
  };
  impl.calls = calls;
  return impl;
}

function makeClient(fetchOpts = {}) {
  const f = fakeFetch(fetchOpts);
  const client = new NxServerClient("https://srv:7001", "admin", "pw", { fetchImpl: f });
  return { client, f };
}

// ---------------------------------------------------------------------------
// login() — uses the latest /rest/v4
// ---------------------------------------------------------------------------

test("login posts credentials to v4 and stores the token", async () => {
  const { client, f } = makeClient({ post: makeResponse({ json: { token: "abc123" } }) });
  const token = await client.login();
  assert.equal(token, "abc123");
  assert.equal(f.calls.post.url, "https://srv:7001/rest/v4/login/sessions");
  assert.deepEqual(f.calls.post.body, { username: "admin", password: "pw", setCookie: false });
});

test("login unauthorized raises AuthError", async () => {
  const { client } = makeClient({ post: makeResponse({ status: 401, text: "bad" }) });
  await assert.rejects(() => client.login(), AuthError);
});

test("login without a token raises ApiError", async () => {
  const { client } = makeClient({ post: makeResponse({ json: { nope: 1 } }) });
  await assert.rejects(() => client.login(), ApiError);
});

// ---------------------------------------------------------------------------
// listCameras()
// ---------------------------------------------------------------------------

test("listCameras returns a plain array and hits v4 /devices", async () => {
  const payload = [{ id: "c1", name: "Lobby", status: "Online", model: "Axis" }];
  const { client, f } = makeClient({ get: makeResponse({ json: payload }) });
  client.token = "abc123";
  const cams = await client.listCameras();
  assert.equal(cams[0].name, "Lobby");
  assert.equal(f.calls.get.url, "https://srv:7001/rest/v4/devices");
  assert.equal(f.calls.get.headers.Authorization, "Bearer abc123");
});

test("listCameras unwraps a reply envelope", async () => {
  const { client } = makeClient({ get: makeResponse({ json: { reply: [{ id: "c1", name: "Lobby" }] } }) });
  client.token = "abc123";
  const cams = await client.listCameras();
  assert.equal(cams.length, 1);
  assert.equal(cams[0].name, "Lobby");
});

test("listCameras without login raises", async () => {
  const { client } = makeClient();
  await assert.rejects(() => client.listCameras(), ApiError);
});

// ---------------------------------------------------------------------------
// logout()
// ---------------------------------------------------------------------------

test("logout deletes the v4 session and clears the token", async () => {
  const { client, f } = makeClient({ del: makeResponse({ json: {} }) });
  client.token = "abc123";
  await client.logout();
  assert.equal(f.calls.deleteCalls, 1);
  assert.equal(f.calls.deleteUrl, "https://srv:7001/rest/v4/login/sessions/abc123");
  assert.equal(client.token, null);
});

test("logout without a token is a no-op", async () => {
  const { client, f } = makeClient();
  await client.logout();
  assert.equal(f.calls.deleteCalls, 0);
});

// ---------------------------------------------------------------------------
// config + table
// ---------------------------------------------------------------------------

test("config uses the NX_SERVER_* vars (env beats file)", () => {
  const args = { host: null, user: null, password: null };
  const config = resolveConfig(args, { NX_SERVER_HOST: "https://file:7001" }, { NX_SERVER_HOST: "https://env:7001" });
  assert.equal(config.host, "https://env:7001");
});

test("formatCamerasTable renders rows and the empty case", () => {
  const out = formatCamerasTable([{ id: "c1", name: "Lobby", status: "Online", model: "Axis" }]);
  assert.ok(out.includes("NAME") && out.includes("Lobby"));
  assert.ok(formatCamerasTable([]).includes("No cameras"));
});
